### PR TITLE
pkg: add tweetnacl as a package

### PR DESCRIPTION
**Pull request by kYc0o**
4th attempt for tweetnacl.

This time the repository resides in git, and two very small modifications were done to make it compile with clang (caused some warnings for different size integers comparisons).

### CODE REVIEW

**Comment by kaspar030** [comment]
Well, I think we're getting there.

File: pkg/tweetnacl/Makefile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+PKG_NAME=tweetnacl
+PKG_URL=https://github.com/kYc0o/tweetnacl
+PKG_VERSION=3548e2c0474731123e8e22a8049a945cd227aea6
+
+.PHONY: all
+
+all: git-download
+	@cp Makefile.tweetnacl $(PKG_BUILDDIR)/Makefile
```
Comment (miri64) [inline pkg/tweetnacl/Makefile]: If you now opened your own Repo, why not add a commit to that that does that for you. Should be pretty easy to maintain (as long as tweetnacl doesn't add their own Makefile ;-))

Comment (kYc0o) [inline pkg/tweetnacl/Makefile]: Do you mean adding such Makefile to the repo? I wanted to create the repo as generic as possible...

**Comment by kaspar030** [comment]
please don't merge, I think we need to do quite some things different.

File: pkg/tweetnacl/Makefile (hunk 1 above)
Comment (OlegHahm) [inline pkg/tweetnacl/Makefile]: Why do you think that adding a Makefile would reduce its generic nature?

**Comment by OlegHahm** [comment]
@kaspar030, can you be more verbose? ;-)

File: pkg/tweetnacl/Makefile (hunk 1 above)
Comment (kYc0o) [inline pkg/tweetnacl/Makefile]: I think adding a makefile with explicit RIOT stuff would avoid to use the sources for other purposes than on RIOT, assuming that such purposes would exist...

**Comment by kaspar030** [comment]
I'll use this PR to test the docker image update on Murdock.

**Comment by kaspar030** [comment]
> @kaspar030, can you be more verbose? ;-)

Yes, sorry, I just wanted to prevent too eager merging just because Murdock shows green.

While we have successfully circumvented our static checks to choke on 3rd-party-code, we're not checking our own anymore. Also, I think it is important to very clearly keep track of our changes to the tweetnacl-code.

So I suggest:

- in the pkg folder, create "src/", move my genrand C file and the Makefile there
- in the tweetnacl repository, make a new initial commit with only the LICENSE and the original tweetnacl sources, documenting the source URL, date and md5's of the tweetnacl files in the commit message, so together with the GIT hashes of our changes, all modifications can be tracked
- make commits with our compile fixes (one for each kind of compile error is preferred)
- in the pkg/tweetnacl Makefile, checkout that directory as usual for git packages, then copy over src/*, and build it

**Comment by OlegHahm** [comment]
Seems like a valid approach to me. I guess it would be best to move the tweetnacl repository into the RIOT-OS organization and use pull requests to modify the code there, too.

**Comment by cgundogan** [comment]
> I guess it would be best to move the tweetnacl repository into the RIOT-OS organization and use pull requests to modify the code there, too.

+1 for that

**Comment by kYc0o** [comment]
I addressed Kaspar's comments and modified the git repo. I'm OK to move it to RIOT-OS, can you verify if the actual state is OK?

**Comment by kaspar030** [comment]
please rebase

File: pkg/tweetnacl/Makefile [hunk 2]
```diff
@@ -0,0 +1,14 @@
+PKG_NAME=tweetnacl
+PKG_URL=https://github.com/kYc0o/tweetnacl
+PKG_VERSION=989553bc3f3990ed4eb7a1f601a10002b2e80ff1
+
+.PHONY: all
+
+all: git-download
+	@cp $(PKG_BUILDDIR)/tweetnacl.c $(RIOTBASE)/pkg/tweetnacl/src/
```
Comment (kaspar030) [inline pkg/tweetnacl/Makefile]: Don't build in src!

**Comment by kaspar030** [comment]
> then copy over src/*, and build it

I meant, copy the contents of pkg/tweetnacl/src into the newly checked out folder in bin/pkg/board/tweetnacl, and build there... Otherwise, parallel builds will fail.

**Comment by kYc0o** [comment]
OK I misunderstood, will do it asap.

**Comment by kYc0o** [comment]
Done. Are you OK with the upstream repo? If you do I can move it as is to RIOT-OS.

**Comment by kaspar030** [comment]
> Are you OK with the upstream repo?

Yes!

**Comment by kYc0o** [comment]
OK so I think now everything is alright. Anyone cares to test it and ACK it?

File: pkg/tweetnacl/Makefile [hunk 3]
```diff
@@ -0,0 +1,11 @@
+PKG_NAME=tweetnacl
+PKG_URL=https://github.com/RIOT-OS/tweetnacl
+PKG_VERSION=7ea05c7098a16c87fa66e9166ce301666f3f2623
```
Comment (OlegHahm) [inline pkg/tweetnacl/Makefile]: Can you add the license information (just introduced in #6301), please?

File: tests/pkg_tweetnacl/Makefile [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# name of your application
```
Comment (OlegHahm) [inline tests/pkg_tweetnacl/Makefile]: I think you can remove all the generic comments from the Makefile.

File: tests/pkg_tweetnacl/Makefile [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+# If no BOARD is found in the environment, use this default:
+BOARD ?= native
+
+# This has to be the absolute path to the RIOT base directory:
+RIOTBASE ?= $(CURDIR)/../..
+
+# Comment this out to disable code in RIOT that does safety checking
+# which is not needed in a production environment but helps in the
+# development process:
+CFLAGS += -DDEVELHELP '-DTHREAD_STACKSIZE_MAIN=(THREAD_STACKSIZE_DEFAULT + 2048)'
```
Comment (OlegHahm) [inline tests/pkg_tweetnacl/Makefile]: Increased stack size is required for tweetnacl? Is this documented somewhere? I would add at least a comment here.

File: pkg/tweetnacl/Makefile (hunk 3 above)
Comment (kYc0o) [inline pkg/tweetnacl/Makefile]: Euh... Yes, of course, but actually there's no official licence for this... it only says "TweetNaCl is a self-contained public-domain C library" so what you would suggest?

File: tests/pkg_tweetnacl/main.c [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+static unsigned char alice_pk[crypto_box_PUBLICKEYBYTES];
+static unsigned char alice_sk[crypto_box_SECRETKEYBYTES];
+
+static unsigned char bob_pk[crypto_box_PUBLICKEYBYTES];
+static unsigned char bob_sk[crypto_box_SECRETKEYBYTES];
+
+static const char message[] = "0123456789abcdef";
+
+/* 17 == strlen(message) + 1 for null byte */
+#define MLEN (17U + crypto_box_ZEROBYTES)
```
Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: `crypto_box_ZEROBYTES` equals 1?

File: tests/pkg_tweetnacl/Makefile (hunk 5 above)
Comment (OlegHahm) [inline tests/pkg_tweetnacl/Makefile]: No `DEVELHELP` for the tests, I would say.

File: tests/pkg_tweetnacl/Makefile [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+# Change this to 0 show compiler invocation lines by default:
+QUIET ?= 1
+
+USEMODULE += ps
```
Comment (OlegHahm) [inline tests/pkg_tweetnacl/Makefile]: Why `ps`?

File: pkg/tweetnacl/Makefile (hunk 3 above)
Comment (OlegHahm) [inline pkg/tweetnacl/Makefile]: PD

Comment (OlegHahm) [inline pkg/tweetnacl/Makefile]: As in https://github.com/RIOT-OS/RIOT/blob/master/pkg/tlsf/Makefile

File: tests/pkg_tweetnacl/main.c (hunk 6 above)
Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: No, `crypto_box_ZEROBYTES` is 32, the comment means that the "magic number" 17 is equals to `strlen(message)` (which is 16) plus the null byte.

File: tests/pkg_tweetnacl/Makefile (hunk 7 above)
Comment (kYc0o) [inline tests/pkg_tweetnacl/Makefile]: Kaspar prints `ps()` at the end of the example, maybe to check that everything is alright? @kaspar030

File: tests/pkg_tweetnacl/main.c (hunk 6 above)
Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: So why not just `#define MLEN (strlen(message) + 1 + crypto_box_ZEROBYTES)`?

File: tests/pkg_tweetnacl/main.c [hunk 8]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2015 Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     examples
```
Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: nope

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: oops...

File: tests/pkg_tweetnacl/Makefile (hunk 5 above)
Comment (kYc0o) [inline tests/pkg_tweetnacl/Makefile]: Done.

**Comment by OlegHahm** [comment]
Do you care to provide a simple (p)expect script for the test case?

**Comment by OlegHahm** [comment]
Apart from the open comments, I'm fine. Tested on native and samr21-xpro.

File: tests/pkg_tweetnacl/main.c (hunk 6 above)
Comment (miri64) [inline tests/pkg_tweetnacl/main.c]: @OlegHahm that isn't evaluated statically. How about `(sizeof(message) + crypto_box_ZEROBYTES)`?

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: At the end I think such changes just make more complex this example... I'd vote to leave it as is, unless you really think the change is strictly necessary.

**Comment by kYc0o** [comment]
All comments addressed.

File: tests/pkg_tweetnacl/main.c (hunk 6 above)
Comment (miri64) [inline tests/pkg_tweetnacl/main.c]: @kYc0o my proposal would make it less complex IMHO, especially considering maintainability. If someone changes `message`, they would also be needed to change this magic number which wouldn't be the case with my proposal ;-)

Comment (miri64) [inline tests/pkg_tweetnacl/main.c]: (`sizeof()` is evaluated at compile-time FYI)

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: OK it works with `sizeof()`. I'll change it.

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: Done.

Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: Oops, sorry, I meant sizeof.

**Comment by kYc0o** [comment]
Alles gut?

File: pkg/tweetnacl/src/Makefile [hunk 9]
```diff
@@ -0,0 +1,5 @@
+MODULE=tweetnacl
+
+SRC := randombytes.c tweetnacl.c
```
Comment (kaspar030) [inline pkg/tweetnacl/src/Makefile]: This line should not be needed.

File: tests/pkg_tweetnacl/Makefile (hunk 7 above)
Comment (kaspar030) [inline tests/pkg_tweetnacl/Makefile]: Yes, as tweetnacl uses a lot of stack, but we don't know yet how much on which platforms...
I propose keeping it in until we know more.

File: tests/pkg_tweetnacl/main.c [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+#include <stdio.h>
+#include <string.h>
+
+#include <tweetnacl.h>
+#include "ps.h"
+
+static const char message[] = "0123456789abcdef";
+
+#define MLEN (sizeof(message) + 1 + crypto_box_ZEROBYTES)
```
Comment (kaspar030) [inline tests/pkg_tweetnacl/main.c]: Const char includes the zero byte when counted with sizeof()

File: tests/pkg_tweetnacl/tests/01-run.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+
+def testfunc(child):
+    child.expect('Creating keypair ALICE...\r\n')
+    child.expect('Done.\r\n')
+    child.expect('Creating keypair BOB...\r\n')
+    child.expect('Done.\r\n')
+    child.expect('Encrypting using pk_bob...\r\n')
+    child.expect('Decrypting...\r\n')
+    child.expect('Decryption successful. content: "0123456789abcdef"\r\n')
+    child.expect('    pid | state    Q | pri \r\n')
```
Comment (kaspar030) [inline tests/pkg_tweetnacl/tests/01-run.py]: Now that I see this, ps doesn't make sense unless develhelp is activated. So remove ps altogether?

**Comment by kaspar030** [comment]
After all this time we have to do it right. ;)

File: tests/pkg_tweetnacl/main.c (hunk 10 above)
Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: Will correct.

File: tests/pkg_tweetnacl/tests/01-run.py (hunk 11 above)
Comment (kYc0o) [inline tests/pkg_tweetnacl/tests/01-run.py]: Uhm... yeah we can remove it. However I can maybe put some more information about this in a README.md, what do you think?

**Comment by kaspar030** [comment]
> Uhm... yeah we can remove it. However I can maybe put some more information about this in a README.md, what do you think?

Yeah! state that tweetnacl needs lots (>2k) of stack space!

**Comment by kYc0o** [comment]
All comments addressed.

File: tests/pkg_tweetnacl/main.c [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+static const unsigned char n[crypto_box_NONCEBYTES];
+
+static unsigned char result[MLEN];
+
+int main(void)
+{
+    int res;
+
+    printf("Creating keypair ALICE...\n");
+    crypto_box_keypair(alice_pk, alice_sk);
```
Comment (miri64) [inline tests/pkg_tweetnacl/main.c]: Sorry for reining in so late, but unless you are using an external library (e.g. [PyNaCl]) these tests would make much more sense as unittests. See #6286 for reasoning.

[PyNaCl]: https://pypi.python.org/pypi/PyNaCl/

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: Uhm... I'm trying to understand what kind of unittests I should write, and what we gain with it.
Actually, to be honest, I think cover with unittests, or even a "normal" test, on cryptographic code is kind of hard task. The test here is just to provide some insights on how to use the library (actually for the use I make of it I used the functions described on this test), and I don't think it aims to actually test the library.

On the other hand, the code is completely independent of RIOT code, except for the random functions which use the module random on RIOT. The code used for the random function is also independent of the library.

In summary, I think any change on the code on the RIOT side will not affect at all this test (e.g. make it fail), so the only source of failure can be upstream changes, which I would suppose should be tested by the upstream CI (if any).

Maybe I'm misunderstanding the purpose of unittests? I see some unittests that cover some crypto libraries, but these libraries are included in the internal code of RIOT, so I see the impact.

Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: I don't think it makes much sense to test third-party code, i.e. adding more tests here, but just transforming the current test into a unittest would provide automatic test execution via the CI.

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: And what's the purpose of such test? I can imagine that it's supposed to be run by the CI on each PR to check if it's not breaking any functionality on the use of this library, but as I stated before, the library does not depend at all on any RIOT code, so no PR on the RIOT side can break it.

Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: > And what's the purpose of such test? 

Exactly the same purpose as the current test.

> but as I stated before, the library does not depend at all on any RIOT code, so no PR on the RIOT side can break it.

What, for instance, if we decide to change the signature for `random_uint32()` or completely drop the function?

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: >>And what's the purpose of such test?

>Exactly the same purpose as the current test.

A unittest to know how to use a library?

>What, for instance, if we decide to change the signature for random_uint32() or completely drop the function?

Ok for that part, but I still think is too much only for testing one function.

I thought the aim of rework this PR was to avoid more charge to our CI, but if it worths, I'll do it.

Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: > A unittest to know how to use a library?

This is a weird purpose for a test (unit or any other) anyway.

Comment (miri64) [inline tests/pkg_tweetnacl/main.c]: The alternative could be to use the independent implementation option ;-)

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: >>A unittest to know how to use a library?

>This is a weird purpose for a test (unit or any other) anyway.

Yeah I know, but until now we don't allow to put an example for each functionality we have in RIOT...

Comment (kYc0o) [inline tests/pkg_tweetnacl/main.c]: I just noted that #5687 already provided some unittests. I adapted it to this PR and pushed it.

**Comment by kYc0o** [comment]
>miri64 approved these changes 6 minutes ago

Yaaaay! OK to squash?

**Comment by miri64** [comment]
Was just about my change requests. The rest I leave to @OlegHahm and @kaspar030.

File: tests/pkg_tweetnacl/main.c [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+ * @author      Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <tweetnacl.h>
+#include "ps.h"
```
Comment (OlegHahm) [inline tests/pkg_tweetnacl/main.c]: remove

File: tests/pkg_tweetnacl/tests/01-run.py (hunk 11 above)
Comment (OlegHahm) [inline tests/pkg_tweetnacl/tests/01-run.py]: ps was removed

File: tests/pkg_tweetnacl/Makefile [hunk 14]
```diff
@@ -0,0 +1,25 @@
+APPLICATION = tweetnacl_test
+
+# If no BOARD is found in the environment, use this default:
```
Comment (OlegHahm) [inline tests/pkg_tweetnacl/Makefile]: Personally, I would remove all the comments except the tweetnacl specific ones here, but that's a matter of taste.

**Comment by kYc0o** [comment]
@OlegHahm comments addressed

**Comment by kYc0o** [comment]
Squashed. Waiting for Murdock while @kaspar030 approves this PR.

**Comment by kaspar030** [comment]
> Squashed. Waiting for Murdock while @kaspar030 approves this PR.

I know this is getting ridiculous, but now there's both a "test" and a
unittest?

Also, the unittest was based on a very old version of the "test"
application which was hastily adapted from code written for Linux, doing
a lot of unneccessary stuff.. "struct identity" ...

Please drop either the unittest or the test application.

**Comment by kYc0o** [comment]
I had a discussion with @OlegHahm about getting the CI checking for this in case the random library gets changed or dropped, so I'd vote to remove the test and leave the unittest. Then you suggest that I should remove such "unnecessary" code?

**Comment by kaspar030** [comment]
> I had a discussion with @OlegHahm about getting the CI checking for this in case the random library gets changed or dropped, so I'd vote to remove the test and leave the unittest. Then you suggest that I should remove such "unnecessary" code?

Well, if you look at the implementation of the unittest, you'll see that
it uses a lot more stack memory than the test.

**Comment by kYc0o** [comment]
>Well, if you look at the implementation of the unittest, you'll see that it uses a lot more stack memory than the test.

Well but this unittest is intended only for the CI right? So I don't think that Murdock will complain about a huge stack usage (I tested in native and works OK).

**Comment by kaspar030** [comment]
> Well but this unittest is intended only for the CI right? So I don't think that Murdock will complain about a huge stack usage (I tested in native and works OK).

Just because the CI *currently* only executes on native doesn't mean we
can be wasteful.

File: tests/unittests/tests-tweetnacl/tests-tweetnacl.c [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+    memset(result, '\0', sizeof(result));
+
+    /* Decrypting... */
+    printf("Decrypting...\n");
+    res = crypto_box_open(result, c, MLEN, n, alice_pk, bob_sk);
+
+    TEST_ASSERT_EQUAL_INT(0, res);
+
+    if (res) {
```
Comment (kaspar030) [inline tests/unittests/tests-tweetnacl/tests-tweetnacl.c]: Could you replace the printed output with a comparison between message and what was decrypted? You'll need a second buffer...

Comment (kYc0o) [inline tests/unittests/tests-tweetnacl/tests-tweetnacl.c]: Done.

**Comment by kaspar030** [comment]
Nice. Please squash!

**Comment by kYc0o** [comment]
Squashed.

**Comment by kYc0o** [comment]
Finally!! Go!! 😃

**Comment by kaspar030** [comment]
@kYc0o Thanks for putting in all the work!

IMHO this PR is a nice example of the extra time needed to make a usable PR into one that passes the strong eyes of our reviewers.

**Comment by kYc0o** [comment]
>IMHO this PR is a nice example of the extra time needed to make a usable PR into one that passes the strong eyes of our reviewers.

Indeed :rofl: but it was a pleasure! Thanks to you, most of the work was done before on your PR 😉
